server/quoridor/quoridor.go

```diff
@@ -0,0 +1,444 @@
+			p.Barriers = barriersForPlayer
+			game.Players[playerNumber] = p
+		} else {
+			playerPawn := Piece{
+				Position: startingPositions[playerNumber],
+				Owner:    playerNumber,
+				Type:     Pawn,
+			}
+			// Create a new player with barrier count, starting position, etc.
+			game.Players[playerNumber] = &Player{
+				Barriers:   barriersForPlayer,
+				PlayerId:   id,
+				PlayerName: name,
+				Pawn:       playerPawn,
+			}
+			// Add pawn to board
+			game.Board[playerPawn.Position] = playerPawn
+			return playerNumber, nil
+		}
+	}
+	return -1, errors.New("no open player positions in this game")
+}
+
+// Advances the turn to the next player.
+func (game *Game) NextTurn() {
+	next := int(game.CurrentTurn+1) % len(game.Players)
+	game.CurrentTurn = PlayerPosition(next)
+}
+
+func (game *Game) IsGameOver() bool {
+	return !game.EndDate.IsZero()
+}
+
+func (game *Game) MaybeReturnWinnerPlayerPosition() PlayerPosition {
+	for position, player := range game.Players {
+		winningPosition := winningPositions[position]
+		if player.Pawn.Position.Y == winningPosition.Y ||
+			player.Pawn.Position.X == winningPosition.X {
+			return position
+		}
+	}
+	return -1
+}
+
+// Deep copy of a game. Will copy all Players, Pieces, and Board.
+func (game *Game) Copy() Game {
+	newGame := Game{
+		Id:          game.Id,
+		CurrentTurn: game.CurrentTurn,
+		StartDate:   game.StartDate,
+		EndDate:     game.EndDate,
+		Winner:      game.Winner,
+		Name:        game.Name,
+	}
+
+	newGame.Board = make(Board)
+	for pos, piece := range game.Board {
+		newPiece := piece
+		newGame.Board[pos] = newPiece
+	}
+
+	newGame.Players = make(map[PlayerPosition]*Player)
+	for pos, player := range game.Players {
+		newGame.Players[pos] = &Player{
+			Barriers:   player.Barriers,
+			Pawn:       player.Pawn,
+			PlayerId:   player.PlayerId,
+			PlayerName: player.PlayerName,
+		}
+	}
+	return newGame
+}
+
+// Starts a game by setting the StartDate to the current instant of time. Returns an error if there aren't enough
+// players, or the game has already started.
+func (game *Game) StartGame() error {
+	if !(len(game.Players) == 2 || len(game.Players) == 4) {
+		return errors.New(fmt.Sprintf("can't start game, wrong number of players (%d)", len(game.Players)))
+	}
+	if !game.StartDate.IsZero() {
+		return errors.New(fmt.Sprintf("game already started"))
+	}
+	game.StartDate = time.Now()
+	return nil
+}
+
+// Initialize with default values
+func NewGame(id uuid.UUID, name string) (*Game, error) {
+	if id == uuid.Nil {
+		return nil, errors.New("unable to create game, need valid id")
+	}
+	if name == "" {
+		return nil, errors.New("unable to create game, need non-empty name")
+	}
+	return &Game{
+		Board:       make(map[Position]Piece),
+		Players:     make(map[PlayerPosition]*Player),
+		CurrentTurn: PlayerOne,
+		Id:          id,
+		Name:        name,
+		Winner:      -1,
+	}, nil
+}
+
+// Return the state of the Board after the move, of the current board if an error
```

Comment:
this comment seems wrong.

Reply:
Definitely - I'm refactoring these methods to only return an error - since the method mutates the Game.